### PR TITLE
AW-163: Edit objects "behind" a transparent shape

### PR DESCRIPTION
**Pull request by matt-aha**
- prioritize selecting objects behind transparent shapes.

In order to support [this feature](https://big.aha.io/features/AW-163), we need to change how `_searchPossibleTargets` prioritizes targets.

This ends up being a fairly simple change to the target search loop. If we encounter an object that isn't completely opaque, we will save it for later and continue searching for targets inside the bounds of that slightly transparent object. I don't think this will have any adverse affects, since the function will always return at least the original slightly transparent object.

Here's how it works in practice:

https://github.com/user-attachments/assets/5943237c-4810-420b-aac8-a099a3a30194

## Testing
The easiest way to test this is to checkout this branch in the `fabric.js` repo locally and point to it in `packages/whiteboard` `package.json` directly (e.g., `fabric: ../../../fabric.js`). You have to `pnpm install` after changing this.

### CODE REVIEW

**Review by gregplaysguitar — COMMENTED** [review]
Haven't had time to review this properly yet, but what happens if you have an object with opacity in front of a solid object, but entirely contained by it? Does this change make it impossible to select the frontmost object?

image

**Comment by matt-aha** [comment]
@gregplaysguitar yeah, good catch.

**Comment by matt-aha** [comment]
@gregplaysguitar I've update the logic to prioritize targets within a transparent target only if they are contained within the transparent target. I think that makes sense, since the only way to click the sub-target would be through the transparent target.

Here's how it works now:

https://github.com/user-attachments/assets/a518cb28-3da4-4d83-a277-8bbde4ac83d7

File: dist/fabric.js [hunk 1]
```diff
@@ -12606,6 +12622,13 @@ fabric.PatternBrush = fabric.util.createClass(fabric.PencilBrush, /** @lends fab
         this._activeObject.clearContextTop();
       }
       fabric.StaticCanvas.prototype.setViewportTransform.call(this, vpt);
+    },
+
+    // Added by aha. Checks if an object is at least partially trasnparent, in
+    // order to prioritize targets behind this object (.e.g, text inside and
+    // behind a slightly transparent square)
+    _isMaybeTransparent: function (object) {
+      return object.fill === 'transparent' || object.opacity < 1;
```
Comment (gregplaysguitar) [inline dist/fabric.js:12631]: Should this include `fill === null` too? I know we set it to 'transparent' but if it were somehow null that'd render the same as transparent?

```suggestion
      return !object.fill || object.fill === 'transparent' || object.opacity < 1;
```

Comment (matt-aha) [reply dist/fabric.js:12631]: Certain objects have `fill: null` that wouldn't be considered transparent, e.g. `SpriteIcon`.

I think `null` is generally used to indicate that the property is no longer used or not being used by an object, so I don't know if we want to key against it in this case.

**Review by gregplaysguitar — APPROVED** [review]
Nice solution

**Pull request merged**